crates/bevy_math/src/primitives/dim2.rs

```diff
@@ -0,0 +1,181 @@
+    ///
+    /// Panics if point1 == point2
+    pub fn from_points(point1: Vec2, point2: Vec2) -> (Self, Vec2) {
+        let diff = point2 - point1;
+        let length = diff.length();
+        (
+            Self::new(Direction2d::from_normalized(diff / length), length),
+            (point1 + point2) / 2.,
+        )
+    }
+
+    /// Get the position of the first point on the line segment
+    pub fn point1(&self) -> Vec2 {
+        *self.direction * -self.half_length
+    }
+
+    /// Get the position of the second point on the line segment
+    pub fn point2(&self) -> Vec2 {
+        *self.direction * self.half_length
+    }
+}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without generics: [`BoxedPolyline2d`]
+#[derive(Clone, Debug)]
+pub struct Polyline2d<const N: usize> {
+    /// The vertices of the polyline
+    pub vertices: [Vec2; N],
+}
+impl<const N: usize> Primitive2d for Polyline2d<N> {}
+
+/// A series of connected line segments in 2D space.
+///
+/// For a version without alloc: [`Polyline2d`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolyline2d {
+    /// The vertices of the polyline
+    pub vertices: Box<[Vec2]>,
+}
+impl Primitive2d for BoxedPolyline2d {}
+
+/// A triangle in 2D space
+#[derive(Clone, Debug)]
+pub struct Triangle2d {
+    /// The vertices of the triangle
+    pub vertices: [Vec2; 3],
+}
+impl Primitive2d for Triangle2d {}
+
+/// A rectangle primitive
+#[doc(alias = "Quad")]
+#[derive(Clone, Copy, Debug)]
+pub struct Rectangle {
+    /// The half width of the rectangle
+    pub half_width: f32,
+    /// The half height of the rectangle
+    pub half_height: f32,
+}
+impl Primitive2d for Rectangle {}
+
+impl Rectangle {
+    /// Create a Rectangle from a full width and height
+    pub fn new(width: f32, height: f32) -> Self {
+        Self::from_size(Vec2::new(width, height))
+    }
+
+    /// Create a Rectangle from the full size of a rectangle
+    pub fn from_size(size: Vec2) -> Self {
+        Self {
+            half_width: size.x / 2.,
+            half_height: size.y / 2.,
+        }
+    }
+}
+
+/// A polygon with N vertices
+/// For a version without generics: [`BoxedPolygon`]
+#[derive(Clone, Debug)]
+pub struct Polygon<const N: usize> {
+    /// The vertices of the polygon
+    pub vertices: [Vec2; N],
+}
+impl<const N: usize> Primitive2d for Polygon<N> {}
+
+/// A polygon with a variable number of vertices
+/// For a version without alloc: [`Polygon`]
+#[derive(Clone, Debug)]
+pub struct BoxedPolygon {
+    /// The vertices of the polygon
+    pub vertices: Box<[Vec2]>,
+}
+impl Primitive2d for BoxedPolygon {}
+
+/// A polygon where all vertices lie on a circle, equally far apart
+#[derive(Clone, Copy, Debug)]
+pub struct RegularPolygon {
+    /// The circumcircle on which all vertices lie
+    pub circumcircle: Circle,
+    /// The number of vertices
+    pub n_vertices: usize,
```

Comment:
Regular polygons are usually defined by the number of sides (*n*-sided polygon), not vertices, although the result is the same. I'm also not a fan of the `n_` prefix, but it's used at least for *n*-sided, so I suppose it might be fine.

Personally I'd just use `sides`:
```suggestion
    /// The number of sides
    pub sides: usize,
```
or `side_count`.

Reply:
If there was a `new` constructor, it could also panic if `sides < 3`